File: 3.x.x/crates-utopia/messaging/src/adapter/sms/msg91.rs
//! PHP `Utopia\Messaging\Adapter\SMS\Msg91` and `MetadataParameter`.

use serde_json::{json, Map, Value};

use super::TYPE;
use crate::adapter::{expect_sms, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::SMS;
use crate::php::ltrim_plus;
use crate::response::{Response, ResponseData};

/// PHP `Adapter\SMS\Msg91\MetadataParameter`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MetadataParameter {
    /// PHP `CLIENT_ID = 'clientId'`.
    ClientId,
    /// PHP `CRQID = 'CRQID'`.
    Crqid,
    /// PHP `UUID = 'UUID'`.
    Uuid,
}

impl MetadataParameter {
    /// PHP enum `value`.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::ClientId => "clientId",
            Self::Crqid => "CRQID",
            Self::Uuid => "UUID",
        }
    }

    /// All cases (PHP `MetadataParameter::cases()`).
    #[must_use]
    pub const fn cases() -> [Self; 3] {
        [Self::ClientId, Self::Crqid, Self::Uuid]
    }
}

/// Validate Msg91 CRQID/UUID metadata (PHP Msg91 + GEOSMS).
pub(crate) fn validate_tracking_metadata(
    metadata: &Map<String, Value>,
) -> Result<(), MessagingError> {
    for key in [
        MetadataParameter::Crqid.as_str(),
        MetadataParameter::Uuid.as_str(),
    ] {
        let Some(value) = metadata.get(key) else {
            continue;
        };
        let Some(text) = value.as_str() else {
            return Err(MessagingError::invalid_argument(format!(
                "Msg91 {key} metadata must be a string."
            )));
        };
        if text.len() > 80 || !tracking_id_ok(text) {
            return Err(MessagingError::invalid_argument(format!(
                "Msg91 {key} metadata must be 80 characters or less and contain only alphanumeric characters, underscores, dots, or hyphens."
            )));
        }
    }
    Ok(())
}

fn tracking_id_ok(value: &str) -> bool {
    !value.is_empty()
        && value
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '.' || c == '-')
}

/// PHP `Adapter\SMS\Msg91`.
#[derive(Debug)]
pub struct Msg91 {
    base: AdapterBase,
    sender_id: String,
    auth_key: String,
    template_id: String,
}

impl Msg91 {
    /// PHP `__construct($senderId, $authKey, $templateId)`.
    #[must_use]
    pub fn new(
        sender_id: impl Into<String>,
        auth_key: impl Into<String>,
        template_id: impl Into<String>,
    ) -> Self {
        Self {
            base: AdapterBase::default(),
            sender_id: sender_id.into(),
            auth_key: auth_key.into(),
            template_id: template_id.into(),
        }
    }

    fn process_sms(&self, message: &SMS) -> Result<ResponseData, MessagingError> {
        let allowed: Vec<&str> = MetadataParameter::cases()
            .iter()
            .map(|p| p.as_str())
            .collect();
        let mut metadata = Map::new();
        if let Some(raw) = message.get_metadata() {
            for (key, value) in raw {
                if allowed.contains(&key.as_str()) {
                    metadata.insert(key.clone(), value.clone());
                }
            }
        }
        for (key, value) in &metadata {
            if !value.is_string() {
                return Err(MessagingError::invalid_argument(format!(
                    "Msg91 {key} metadata must be a string."
                )));
            }
        }
        validate_tracking_metadata(&metadata)?;

        let recipients: Vec<Value> = message
            .get_to()
            .iter()
            .map(|recipient| {
                json!({
                    "mobiles": ltrim_plus(recipient),
                    "content": message.get_content(),
                    "otp": message.get_content(),
                })
            })
            .collect();

        let mut body = json!({
            "sender": self.sender_id,
            "template_id": self.template_id,
            "recipients": recipients,
        });
        if let Some(obj) = body.as_object_mut() {
            for (key, value) in metadata {
                obj.insert(key, value);
            }
        }

        let result = self.request_default(
            "POST",
            "https://api.msg91.com/api/v5/flow/",
            &[
                "Content-Type: application/json".into(),
                format!("Authkey: {}", self.auth_key),
            ],
            Some(body),
        );
        let mut response = Response::new(TYPE);
        if result.status_code == 200 {
            response.set_delivered_to(message.get_to().len() as i64);
            for to in message.get_to() {
                response.add_result(to, "");
            }
        } else {
            for to in message.get_to() {
                response.add_result(to, "Unknown error");
            }
        }
        Ok(response.to_array())
    }
}

impl Adapter for Msg91 {
    fn get_name(&self) -> &'static str {
        "Msg91"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::SMS
    }
    fn get_max_messages_per_request(&self) -> usize {
        100
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(
            self.process_sms(expect_sms(message)?)?,
        ))
    }
}

File: packages/messaging/src/Adapter/SMS/Msg91.php
<?php

namespace Utopia\Messaging\Adapter\SMS;

use Utopia\Messaging\Adapter\SMS as SMSAdapter;
use Utopia\Messaging\Adapter\SMS\Msg91\MetadataParameter;
use Utopia\Messaging\Messages\SMS as SMSMessage;
use Utopia\Messaging\Response;

// Reference Material
// https://docs.msg91.com/p/tf9GTextN/e/7WESqQ4RLu/MSG91

class Msg91 extends SMSAdapter
{
    protected const NAME = 'Msg91';

    /**
     * @param  string  $senderId Msg91 Sender ID
     * @param  string  $authKey Msg91 Auth Key
     * @param  string  $templateId Msg91 Template ID
     */
    public function __construct(
        private readonly string $senderId,
        private readonly string $authKey,
        private readonly string $templateId,
    ) {
        parent::__construct();
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getMaxMessagesPerRequest(): int
    {
        // TODO: Find real limit
        return 100;
    }

    /**
     * {@inheritdoc}
     */
    protected function process(SMSMessage $message): array
    {
        $metadata = $message->getMetadata() ?? [];
        $metadata = array_intersect_key($metadata, array_flip(array_column(MetadataParameter::cases(), 'value')));

        foreach ($metadata as $key => $value) {
            if (!\is_string($value)) {
                throw new \InvalidArgumentException("Msg91 {$key} metadata must be a string.");
            }
        }

        foreach ([MetadataParameter::CRQID, MetadataParameter::UUID] as $parameter) {
            $key = $parameter->value;

            if (!\array_key_exists($key, $metadata)) {
                continue;
            }

            if (\strlen($metadata[$key]) > 80 || !preg_match('/^[A-Za-z0-9_.-]+$/', $metadata[$key])) {
                throw new \InvalidArgumentException("Msg91 {$key} metadata must be 80 characters or less and contain only alphanumeric characters, underscores, dots, or hyphens.");
            }
        }

        $recipients = [];
        foreach ($message->getTo() as $recipient) {
            $recipients[] = [
                'mobiles' => ltrim($recipient, '+'),
                'content' => $message->getContent(),
                'otp' => $message->getContent(),
            ];
        }

        $body = [
            'sender' => $this->senderId,
            'template_id' => $this->templateId,
            'recipients' => $recipients,
        ];

        foreach ($metadata as $key => $value) {
            $body[$key] = $value;
        }

        $response = new Response($this->getType());
        $result = $this->request(
            method: 'POST',
            url: 'https://api.msg91.com/api/v5/flow/',
            headers: [
                'Content-Type: application/json',
                'Authkey: ' . $this->authKey,
            ],
            body: $body,
        );

        if ($result['statusCode'] === 200) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to);
            }
        } else {
            foreach ($message->getTo() as $to) {
                $response->addResult($to, 'Unknown error');
            }
        }

        return $response->toArray();
    }
}
